pandas/core/arrays/masked.py

```diff
@@ -217,6 +218,18 @@ def copy(self: BaseMaskedArrayT) -> BaseMaskedArrayT:
         mask = mask.copy()
         return type(self)(data, mask, copy=False)
 
+    @doc(ExtensionArray.factorize)
+    def factorize(self, na_sentinel: int = -1) -> Tuple[np.ndarray, ExtensionArray]:
+        arr = self._data
+        mask = self._mask
+
+        codes, uniques = _factorize_array(arr, na_sentinel=na_sentinel, mask=mask)
+
+        # the hashtables don't handle all different types of bits
+        uniques = uniques.astype(self.dtype.numpy_dtype, copy=False)
```

Comment:
I get a "pandas/core/arrays/masked.py:229: error: "ExtensionDtype" has no attribute "numpy_dtype"" mypy failure

cc @simonjayhawkins @WillAyd how can I solve / silence this? The `numpy_dtype` attribute is commong for Int/BoolDtype (so I can safely use it), but not for general ExtensionDtype.

Reply:
Hmm yea so I guess complaining because as far as this class is defined, the return type of `self.dtype` is an ExtensionDtype (as defined in `ExtensionArray`)

I guess it comes back to the class design; if we have something else that inherits from BaseMaskedArray it could fail at runtime without if it isn't constructed to return a dtype from `self.dtype` that has a `numpy_dtype` attribute, which is a little hefty on the implicitness I guess

Reply:
So the subclasses IntegerArray and BooleanArray have a correctly types `dtype` property. But this method above is defined in their parent class ..

In principle I could add a `dtype` property

```
@property
def dtype(self) -> Union["IntegerDtype", "BooleanDtype"]:
    pass
 ```

in the BaseMaskedArray class to solve this, I suppose? 
But that is also kind of ugly, as the parent class shouldn't really know about its subclasses ..

Reply:
Right; I think it's going to be tough to make this work with mypy if we implicitly enforce that subclasses make `dtype.numpy_dtype` available

What does the comment directly preceding it refer to? Perhaps there is a way to do this without breaking the currently implied subclass requirements?

Reply:
Can you override the type signature of IntegerArray.dtype to be `IntegerDtype` and BolleanArray.dtype to be BooleanDtype?

Reply:
Oh that's already the case... How about a `MaskedArrayDtype` that subclasses ExtensionDtype but has a `numpy_dtype` property?

and then BooleanDtype and IntegerDtype inherit from that rather than ExtensionDtype.

Reply:
Is there a way to just disable mypy on this line?

> What does the comment directly preceding it refer to? Perhaps there is a way to do this without breaking the currently implied subclass requirements?

The hashtable is only implemented for int64. So if you have an int32 array, the unique values coming out of `_factorize_array` are int64, and need to be casted back to int32 (as the uniques returned from this method should be using the original dtype). So for this casting, I need to have access to the dtype's equivalent numpy dtype, which is avalaible as the `numpy_dtype` attribute. 

I could do this differently by eg building up a mapping of EADtypes -> numpy dtypes and looking it up from there instead of using the attribute, but that would just be introducing more complex workarounds to just to satisfy mypy.

Reply:
> How about a MaskedArrayDtype that subclasses ExtensionDtype but has a numpy_dtype property?

Yes, that's probably the cleanest solution architecturally

Reply:
> How about a MaskedArrayDtype that subclasses ExtensionDtype but has a numpy_dtype property?

Yes, that's probably the cleanest solution architecturally. 
But the `dtype` attribute on BaseMaskedArray would still only be a dummy property just to provide typing, since it is overwritten in both subclasses.

Reply:
> But the dtype attribute on BaseMaskedArray would still only be a dummy property just to provide typing

I don't think you'll need to change anything on the array side. The dtypes will inherit from MaskedExtensionDtype, so mypy should know that `integer_array.dtype.numpy_type` is valid.

Or we add `numpy_dtype` to the `ExtensionDtype` API :)

Reply:
> I don't think you'll need to change anything on the array side. The dtypes will inherit from MaskedExtensionDtype, so mypy should know that integer_array.dtype.numpy_type is valid.

No, since mypy thinks `self.dtype` is an ExtensionDtype, so having IntegerDtype/BooleanDtype inherit from a MaskedDtype that defines this attribute will no help.

So either we would indeed need to add `numpy_dtype` to the ExtensionDtype API, or I need to add a dummy `dtype` property on BaseMaskedArray to be able to type it as MaskedDtype.